From 9845a0601f84796f32cc8d07e120429b69b80896 Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Mon, 29 May 2023 18:34:07 +0800
Subject: [PATCH 56/68] configs: arm64: add harpoon virtio_net cells for imx93

Use a standalone cell in order to avoid to disabling all enet ports.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
---
 configs/arm64/imx9-harpoon-config-virtio.h    | 53 +++++++++++++++++++
 configs/arm64/imx93-harpoon-freertos-virtio.c | 22 ++++++++
 configs/arm64/imx93-harpoon-rtos.h            | 26 +++++++++
 3 files changed, 101 insertions(+)
 create mode 100644 configs/arm64/imx9-harpoon-config-virtio.h
 create mode 100644 configs/arm64/imx93-harpoon-freertos-virtio.c

diff --git a/configs/arm64/imx9-harpoon-config-virtio.h b/configs/arm64/imx9-harpoon-config-virtio.h
new file mode 100644
index 00000000..a4d58c24
--- /dev/null
+++ b/configs/arm64/imx9-harpoon-config-virtio.h
@@ -0,0 +1,53 @@
+/*
+ * i.MX93 family targets - Virtio_net config
+ *
+ * Copyright 2023 NXP
+ *
+ * Authors:
+ *  Jiafei Pan <Jiafei.Pan@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#define CONFIG_INMATE_BASE		RTOS_INMATE_BASE
+
+/* Memory & peripherals */
+#define CONFIG_INMATE_REGIONS                   \
+	RTOS_COMMON_MEMORY_REGIONS,                               \
+	RTOS_MEMRAM_MEMORY_REGIONS,                               \
+	RTOS_VIRTIO_MEMORY_REGIONS
+
+/* GIC */
+#define CONFIG_INMATE_IRQCHIPS_ADDR                   \
+	CONFIG_RTOS_IRQCHIPS_ADDR
+#define CONFIG_INMATE_IRQCHIPS 	                                         \
+	IRQCHIP0(CONFIG_INMATE_IRQCHIPS_ADDR,                            \
+		/* interrupts 32..63 */                                  \
+		RTOS_COMMON_IRQCHIP0_BITMAP1 |                           \
+		RTOS_VIRTIO_IRQCHIP0_BITMAP1,                             \
+		/* interrupts 64..95 */                                  \
+		RTOS_COMMON_IRQCHIP0_BITMAP2 |                           \
+		RTOS_VIRTIO_IRQCHIP0_BITMAP2,                             \
+		/* interrupts 96..127 */                                 \
+		RTOS_COMMON_IRQCHIP0_BITMAP3 |                           \
+		RTOS_VIRTIO_IRQCHIP0_BITMAP3,                             \
+		/* interrupts 128..159 */                                \
+		RTOS_COMMON_IRQCHIP0_BITMAP4 |                           \
+		RTOS_VIRTIO_IRQCHIP0_BITMAP4),                            \
+	IRQCHIP1(CONFIG_INMATE_IRQCHIPS_ADDR,                            \
+		/* interrupts 160..191 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP1 |                           \
+		RTOS_VIRTIO_IRQCHIP1_BITMAP1,                             \
+		/* interrupts 192..223 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP2 |                           \
+		RTOS_VIRTIO_IRQCHIP1_BITMAP2,                            \
+		/* interrupts 224..255 */                                \
+		0,                                                       \
+		/* interrupts 256..287 */                                \
+		0)
+
+/* IVSHMEM PCI */
+#define CONFIG_INMATE_PCI_DEVICES                    \
+	RTOS_COMMON_PCI_IVSHMEM
+
diff --git a/configs/arm64/imx93-harpoon-freertos-virtio.c b/configs/arm64/imx93-harpoon-freertos-virtio.c
new file mode 100644
index 00000000..97e9de8a
--- /dev/null
+++ b/configs/arm64/imx93-harpoon-freertos-virtio.c
@@ -0,0 +1,22 @@
+/*
+ * i.MX93 target - FreeRTOS virtio_net config
+ *
+ * Copyright 2023 NXP
+ *
+ * Authors:
+ *  Jiafei Pan <jiafei.pan@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#include "cell-helper.h"
+#include "imx93-harpoon-rtos.h"
+
+#define CONFIG_INMATE_NAME		"freertos"
+#define CONFIG_INMATE_CORE_BITMAP	(0b10)
+
+/* Base, memory & peripherals, GIC, IVSHMEM PCI */
+#include "imx9-harpoon-config-virtio.h"
+
+#include "cell-create.h"
diff --git a/configs/arm64/imx93-harpoon-rtos.h b/configs/arm64/imx93-harpoon-rtos.h
index b45a6d55..c59be13b 100644
--- a/configs/arm64/imx93-harpoon-rtos.h
+++ b/configs/arm64/imx93-harpoon-rtos.h
@@ -15,6 +15,9 @@ typedef enum IRQn {
   LPUART2_IRQn                 = 52,               /**< Low Power UART 2 */
   TPM2_IRQn                    = 69,               /**< Timer PWM Module (TPM) 2 */
   TPM4_IRQn                    = 108,               /**< Timer PWM Module (TPM) 4 */
+  SW_MB_L_IRQn                 = 200,              /**< Unused SPI 168, to be used by generic software mailbox in Linux */
+  SW_MB_R_IRQn                 = 201,              /**< Unused SPI 169, to be used by generic software mailbox in RTOS */
+  ENET_IRQn                    = 213,               /**< ENET1 MAC Interrupt */
 } IRQn_Type;
 
 #define RTOS_COMMON_MEMORY_REGIONS \
@@ -24,6 +27,9 @@ typedef enum IRQn {
 	MEM_REGION_RWS( 0xfd9fc000, 0xfd9fc000, KB(8)),  /* IVSHMEM */              \
 	MEM_REGION_ROS( 0xfd9fe000, 0xfd9fe000, KB(8)),  /* IVSHMEM */              \
 	/* IVSHMEM regions should always be kept on top so PCI_IVS doesn't change */\
+	MMIO_REGION_RWS(0xf8700000, 0xf8700000, MB(64)), /* Virtio DMA buffer */    \
+	MMIO_REGION_RWS(0xfc700000, 0xfc700000, MB(1)),  /* Virtio MMIO */          \
+	MMIO_REGION_RWS(0xfe000000, 0xfe000000, KB(4)),  /* SW MBOX */              \
 	MMIO_REGION_RWS(0x443c0000, 0x443c0000, KB(64)), /* IOMUXC */               \
 	MMIO_REGION_RWS(0x44450000, 0x44450000, KB(64)), /* CCM */                  \
 	MMIO_REGION_ROS(0x44480000, 0x44480000, KB(64)), /* ANA_PLL */              \
@@ -35,6 +41,10 @@ typedef enum IRQn {
 	MEM_REGION_RWXL(0xd0000000, 0xd0000000, MB(16)), /* RAM */    \
 	MEM_REGION_RWX( 0x20480000, 0x20480000, KB(640)) /* OCRAM */
 
+#define RTOS_VIRTIO_MEMORY_REGIONS \
+	MMIO_REGION_RW( 0x42890000, 0x42890000, KB(64)),   /* ENET */               \
+	MEM_REGION_RWS( 0x80000000, 0x80000000, GB(1))   /* data buffer */
+
 /* GIC */
 #define CONFIG_RTOS_IRQCHIPS_ADDR	(0x48000000)
 
@@ -48,6 +58,8 @@ typedef enum IRQn {
 	RTOS_AUDIO_IRQCHIP0_BITMAP1
 #define RTOS_INDUS_IRQCHIP0_BITMAP1 \
 	0
+#define RTOS_VIRTIO_IRQCHIP0_BITMAP1 \
+	0
 
 /* interrupts 64..95 */
 #define RTOS_COMMON_IRQCHIP0_BITMAP2 \
@@ -59,6 +71,8 @@ typedef enum IRQn {
 	RTOS_AUDIO_IRQCHIP0_BITMAP2
 #define RTOS_INDUS_IRQCHIP0_BITMAP2 \
 	0
+#define RTOS_VIRTIO_IRQCHIP0_BITMAP2 \
+	0
 
 /* interrupts 96..127 */
 #define RTOS_COMMON_IRQCHIP0_BITMAP3 \
@@ -70,6 +84,8 @@ typedef enum IRQn {
 	RTOS_AUDIO_IRQCHIP0_BITMAP3
 #define RTOS_INDUS_IRQCHIP0_BITMAP3 \
 	0
+#define RTOS_VIRTIO_IRQCHIP0_BITMAP3 \
+	0
 
 /* interrupts 128..159 */
 #define RTOS_COMMON_IRQCHIP0_BITMAP4 \
@@ -81,8 +97,12 @@ typedef enum IRQn {
 		RTOS_AUDIO_IRQCHIP0_BITMAP4
 #define RTOS_INDUS_IRQCHIP0_BITMAP4 \
 	0
+#define RTOS_VIRTIO_IRQCHIP0_BITMAP4 \
+	0
 
 /* interrupts 160..191 */
+#define RTOS_COMMON_IRQCHIP1_BITMAP1 \
+	0
 #define RTOS_AUDIO_IRQCHIP1_BITMAP1 \
 	0
 #define RTOS_AVB_IRQCHIP1_BITMAP1 \
@@ -90,8 +110,12 @@ typedef enum IRQn {
 	0
 #define RTOS_INDUS_IRQCHIP1_BITMAP1 \
 	0
+#define RTOS_VIRTIO_IRQCHIP1_BITMAP1 \
+	0
 
 /* interrupts 192..223 */
+#define RTOS_COMMON_IRQCHIP1_BITMAP2 \
+	IRQnMAP(SW_MB_R_IRQn)
 #define RTOS_AUDIO_IRQCHIP1_BITMAP2 \
 	0
 #define RTOS_AVB_IRQCHIP1_BITMAP2 \
@@ -99,6 +123,8 @@ typedef enum IRQn {
 	0
 #define RTOS_INDUS_IRQCHIP1_BITMAP2 \
 	0
+#define RTOS_VIRTIO_IRQCHIP1_BITMAP2 \
+	IRQnMAP(ENET_IRQn)
 
 /* IVSHMEM PCI */
 #define	RTOS_COMMON_PCI_IVSHMEM \
-- 
2.34.1

